File: platform/platform-impl/src/com/intellij/ide/actions/DebugLogConfigureAction.java
/*
 * Copyright 2000-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intellij.ide.actions;

import com.intellij.diagnostic.DebugLogManager;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.DumbAwareAction;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.project.ProjectManager;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.ScrollPaneFactory;
import com.intellij.ui.components.JBLabel;
import com.intellij.util.Function;
import com.intellij.util.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.List;

public class DebugLogConfigureAction extends DumbAwareAction {

  @Override
  public void actionPerformed(AnActionEvent e) {
    Project project = e.getProject() == null ? ProjectManager.getInstance().getDefaultProject() : e.getProject();
    DebugLogManager logCustomizer = ServiceManager.getService(DebugLogManager.class);
    DebugLogConfigureDialog dialog = new DebugLogConfigureDialog(project, logCustomizer.getSavedCategories());
    if (dialog.showAndGet()) {
      List<String> categories = dialog.getLogCategories();
      logCustomizer.applyCategories(categories);
      logCustomizer.saveCategories(categories);
    }
  }

  private static class DebugLogConfigureDialog extends DialogWrapper {

    private static final String ALL_POSSIBLE_SEPARATORS = "(\n|,|;)+";
    @NotNull private final JTextArea myTextArea;

    protected DebugLogConfigureDialog(@Nullable Project project, List<String> categories) {
      super(project, false);
      myTextArea = new JTextArea(10, 30);
      myTextArea.setText(StringUtil.join(categories, "\n"));
      setTitle("Custom Debug Log Configuration");
      init();
    }

    @Nullable
    @Override
    protected JComponent createNorthPanel() {
      return new JBLabel("Add log categories separated by new lines");
    }

    @Nullable
    @Override
    protected JComponent createCenterPanel() {
      return ScrollPaneFactory.createScrollPane(myTextArea);
    }

    @Nullable
    @Override
    public JComponent getPreferredFocusedComponent() {
      return myTextArea;
    }

    @NotNull
    public List<String> getLogCategories() {
      return parseCategories(myTextArea.getText());
    }

    @NotNull
    private static List<String> parseCategories(@NotNull String text) {
      return ContainerUtil.mapNotNull(text.split(ALL_POSSIBLE_SEPARATORS), new Function<String, String>() {
        @Override
        public String fun(String s) {
          return StringUtil.isEmptyOrSpaces(s) ? null : s.trim();
        }
      });
    }
  }

}

File: platform/platform-impl/src/com/intellij/ide/actions/DebugLogConfigureAction.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.ide.actions

import com.intellij.diagnostic.logs.DebugLogLevel
import com.intellij.diagnostic.logs.LogCategory
import com.intellij.diagnostic.logs.LogLevelConfigurationManager
import com.intellij.diagnostic.logs.LogLevelConfigurationManager.State
import com.intellij.ide.IdeBundle
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.DumbAwareAction
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.ui.ScrollPaneFactory
import com.intellij.ui.components.JBLabel
import com.intellij.util.ui.JBUI
import com.intellij.xml.util.XmlStringUtil
import javax.swing.JTextArea

internal class DebugLogConfigureAction : DumbAwareAction() {
  override fun actionPerformed(e: AnActionEvent) {
    val project = e.project ?: ProjectManager.getInstance().defaultProject
    val logLevelConfigurationManager = LogLevelConfigurationManager.getInstance()
    val currentState = logLevelConfigurationManager.state

    val dialog = DebugLogConfigureDialog(project, currentState)
    if (dialog.showAndGet()) {
      val dialogState = dialog.getLogCategories()
      logLevelConfigurationManager.setCategories(dialogState)
    }
  }

  override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.BGT
}

private const val TRACE_TOKEN = "trace"
private const val ALL_TOKEN = "all"
private const val SEPARATE_FILE_TOKEN = "separate"
private val ALL_POSSIBLE_SEPARATORS = "[\n,;]+".toRegex()

private class DebugLogConfigureDialog(project: Project, state: State) : DialogWrapper(project, false) {
  private val myTextArea = JTextArea(10, 30)

  init {
    myTextArea.margin = JBUI.insets(2)
    myTextArea.text = state.categories.joinToString("\n") {
      var line = when (it.level) {
        DebugLogLevel.DEBUG -> it.category
        DebugLogLevel.TRACE -> "${it.category}:$TRACE_TOKEN"
        DebugLogLevel.ALL -> "${it.category}:$ALL_TOKEN"
      }
      if (it.category in state.categoriesWithSeparateFiles) {
        line += ":$SEPARATE_FILE_TOKEN"
      }
      line
    }
    title = IdeBundle.message("dialog.title.custom.debug.log.configuration")
    init()
  }

  override fun getDimensionServiceKey() = "#com.intellij.ide.actions.DebugLogConfigureAction"

  override fun createNorthPanel() = JBLabel(
    XmlStringUtil.wrapInHtml(IdeBundle.message("label.enable.debug.level", ":$TRACE_TOKEN", ":$ALL_TOKEN", ":$SEPARATE_FILE_TOKEN")))

  override fun createCenterPanel() = ScrollPaneFactory.createScrollPane(myTextArea)

  override fun getPreferredFocusedComponent() = myTextArea

  fun getLogCategories(): State {
    val logCategories = mutableListOf<LogCategory>()
    val categoriesWithDedicatedFiles = hashSetOf<String>()
    for (line in myTextArea.text.split(ALL_POSSIBLE_SEPARATORS)) {
      val line = line.trim()
      if (line.isNotEmpty()) {
        val split = line.split(':')
        var level = DebugLogLevel.DEBUG
        val category = split[0].trim()
        for (token in split.drop(1)) {
          when (token.lowercase().trim()) {
            TRACE_TOKEN -> level = DebugLogLevel.TRACE
            ALL_TOKEN -> level = DebugLogLevel.ALL
            SEPARATE_FILE_TOKEN -> categoriesWithDedicatedFiles.add(category)
          }
        }
        logCategories.add(LogCategory(category, level))
      }
    }
    return State(logCategories, categoriesWithDedicatedFiles)
  }
}